### PR TITLE
Martin/ios 611 add emotion detail flow

### PR DESCRIPTION
**Pull request by MartinJibber**
Getting the basic emotion detail screen working.

https://user-images.githubusercontent.com/88675954/165187894-a08760fa-106f-4ec6-a9ea-6813151b8d18.mp4

### CODE REVIEW

File: Shared/Flows/Launch/SplashViewController.swift [hunk 1]
```diff
@@ -19,7 +19,7 @@ class SplashViewController: FullScreenViewController, TransitionableViewControll
     
     /// A view to blur out the emotions collection view.
     let blurView = BlurView()
-    private lazy var emotionCollectionView = EmotionCircleCollectionView(cellDiameter: 80)
+    private lazy var emotionCollectionView = EmotionCircleCollectionView(cellDiameter: 100)
```
Comment (MartinJibber) [inline Shared/Flows/Launch/SplashViewController.swift:22]: Making the default circle size on the splash screen bigger.

File: Shared/Flows/Launch/SplashViewController.swift [hunk 2]
```diff
@@ -67,9 +67,12 @@ class SplashViewController: FullScreenViewController, TransitionableViewControll
     private func animateEmotions() async {
         guard !Task.isCancelled else { return }
         
-        guard let emotion = Emotion.allCases.randomElement() else { return }
+        let emotions = Emotion.allCases.random(Int.random(min: 1, max: 4))
+        var emotionsCounts: [Emotion : Int] = [:]
+        for emotion in emotions {
+            emotionsCounts[emotion] = Int.random(min: 1, max: 3)
+        }
```
Comment (MartinJibber) [inline Shared/Flows/Launch/SplashViewController.swift:74]: Showing more circles on the splash screen and randomizing their sizes.

File: Shared/UI/Transitions/Transition Types/TransitionRouter+Blur.swift [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+//  Copyright © 2022 Benjamin Dodgson. All rights reserved.
+//
+
+import UIKit
+
+extension TransitionRouter {
+
+    /// For presentation, blurs the fromVC then fades in the toVC.
+    /// For dismiss, fades out fromVC and unblurs the toVC.
+    func blur(transitionContext: UIViewControllerContextTransitioning) {
```
Comment (MartinJibber) [inline Shared/UI/Transitions/Transition Types/TransitionRouter+Blur.swift:15]: I made a new transition function to blur the background and fade in the presented vc.

File: Shared/UI/Transitions/Transition Types/TransitionRouter+Fade.swift [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import UIKit
+
+extension TransitionRouter {
+
+    /// Fades out the fromVC and then fades in the toVC.
+    func fadeTranstion(transitionContext: UIViewControllerContextTransitioning) {
+        let containerView = transitionContext.containerView
+
+        let isPresenting = self.operation == .push
+        if isPresenting {
```
Comment (MartinJibber) [inline Shared/UI/Transitions/Transition Types/TransitionRouter+Fade.swift:19]: Fixing so bugs related to dismissal in this transition animator.

File: Shared/UI/Transitions/TransitionRouter.swift [hunk 5]
```diff
@@ -8,14 +8,15 @@
...
+
 /// A custom transitions that simultaneously slides the current VC vertically off the screen
 /// and the destination one onto it.
 class TransitionRouter: NSObject, UIViewControllerAnimatedTransitioning {
 
     // The fromVC sets the stage for how it wants to get to the toVC
     private(set) var fromVC: TransitionableViewController
     private(set) var toVC: TransitionableViewController
-    private let operation: UINavigationController.Operation
+    let operation: UINavigationController.Operation
```
Comment (MartinJibber) [inline Shared/UI/Transitions/TransitionRouter.swift:19]: Some of the animation functions need to know which operation is happening.

File: Shared/UI/Views/MessageContentView/MessageContentView.swift [hunk 6]
```diff
@@ -18,6 +18,9 @@ protocol MessageContentDelegate: AnyObject {
     func messageContent(_ content: MessageContentView, didTapEditMessage messageInfo: (ConversationId, MessageId))
     func messageContent(_ content: MessageContentView, didTapAttachmentForMessage messageInfo: (ConversationId, MessageId))
     func messageContent(_ content: MessageContentView, didTapAddEmotionsForMessage messageInfo: (ConversationId, MessageId))
+    func messageContent(_ content: MessageContentView,
+                        didTapEmotion emotion: Emotion,
+                        forMessage messageInfo: (ConversationId, MessageId))
```
Comment (MartinJibber) [inline Shared/UI/Views/MessageContentView/MessageContentView.swift:23]: New message content delegate function.

File: iOS/Flows/Conversation/Emotion Detail/EmotionDetailCollectionViewDataSource.swift [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+//  Copyright © 2022 Benjamin Dodgson. All rights reserved.
+//
+
+import Foundation
+
+typealias EmotionDetailSection = EmotionDetailCollectionViewDataSource.SectionType
+typealias EmotionDetailItem = EmotionDetailCollectionViewDataSource.ItemType
+
+class EmotionDetailCollectionViewDataSource: CollectionViewDataSource<EmotionDetailSection,
+                                             EmotionDetailItem> {
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Emotion Detail/EmotionDetailCollectionViewDataSource.swift:15]: The layout and data source below are similar but simpler than the emotion selection screen.

File: iOS/Flows/Conversation/Emotion Detail/EmotionDetailCollectionViewLayout.swift [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+                                                              trailing: sectionInset)
+
+                // Section
+                let section = NSCollectionLayoutSection(group: group)
+                section.contentInsets = NSDirectionalEdgeInsets(top: sectionInset,
+                                                                leading: 0,
+                                                                bottom: 20,
+                                                                trailing: 0)
+
+                section.orthogonalScrollingBehavior = .groupPagingCentered
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Emotion Detail/EmotionDetailCollectionViewLayout.swift:50]: Paging centered is a nice option to have now.

File: iOS/Flows/Conversation/Emotion Detail/EmotionDetailCoordinator.swift [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+//  EmotionDetailCoordinator.swift
+//  Jibber
+//
+//  Created by Martin Young on 4/25/22.
+//  Copyright © 2022 Benjamin Dodgson. All rights reserved.
+//
+
+import Foundation
+
+class EmotionDetailCoordinator: PresentableCoordinator<Void> {
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Emotion Detail/EmotionDetailCoordinator.swift:11]: Super simple coordinator. Pretty much just presents the emotion details.

File: iOS/Flows/Conversation/Emotion Detail/EmotionDetailViewController.swift [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+        self.delegate = delegate
+
+        var sortedEmotions = emotions
+        // Put the starting emotion at the front
+        if let startingEmotion = startingEmotion {
+            sortedEmotions.remove(object: startingEmotion)
+            sortedEmotions.insert(startingEmotion, at: 0)
+        }
+        self.emotions = sortedEmotions
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Emotion Detail/EmotionDetailViewController.swift:35]: See comment.

File: iOS/Flows/Conversation/Emotions/EmotionContentCell.swift [hunk 11]
```diff
@@ -14,61 +14,70 @@ struct EmotionContentModel: Hashable {
...
+        self.definitionLabel.setSize(withWidth: self.circleView.width * 0.8)
+        self.definitionLabel.centerOnXAndY()
+
+        self.emotionLabel.setSize(withWidth: self.circleView.width)
+
+        self.definitionLabel.top += self.emotionLabel.height
+
         self.emotionLabel.centerOnX()
-        self.emotionLabel.match(.bottom, to: .top, of: self.label, offset: .negative(.standard))
+        self.emotionLabel.match(.bottom, to: .top, of: self.definitionLabel, offset: .negative(.standard))
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Emotions/EmotionContentCell.swift:81]: Some layout adjustments to the text. Now the title and definition are vertically centered as a unit.

File: iOS/Flows/Conversation/Emotions/EmotionContentCell.swift [hunk 12]
```diff
@@ -14,61 +14,70 @@ struct EmotionContentModel: Hashable {
...
         super.layoutSubviews()
-        
-        self.label.setSize(withWidth: self.contentView.width * 0.8)
-        self.label.centerOnXAndY()
-        
-        self.emotionLabel.setSize(withWidth: self.contentView.width)
+
+        self.circleView.squaredSize = min(self.contentView.width, self.contentView.height)
+        self.circleView.centerOnXAndY()
+        self.circleView.makeRound()
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Emotions/EmotionContentCell.swift:71]: Making it circular.

File: iOS/Models/EmotionCategory.swift [hunk 13]
```diff
@@ -95,7 +95,7 @@ enum EmotionCategory: Int, CaseIterable {
         case .compare:
             return [.comparison, .admiration, .reverence, .envy, .jealous, .resentment, .schadenfreude, .freudenfreude]
         case .unplanned:
-            return [.bored, .dissapointment, .expecting, .regret, .discouraged, .resigned, .frustrated]
+            return [.bored, .disappointment, .expecting, .regret, .discouraged, .resigned, .frustrated]
```
Comment (MartinJibber) [inline iOS/Models/EmotionCategory.swift:98]: Typo

**Review by Benjamindodgson — APPROVED** [review]
LOVE

**Pull request merged**